File: android/app/src/main/java/com/smartexpense/android/data/repository/NotificationRepositoryImpl.java
package com.smartexpense.android.data.repository;

import com.smartexpense.android.data.remote.api.NotificationApi;
import com.smartexpense.android.data.remote.dto.response.NotificationResponseDto;
import com.smartexpense.android.domain.repository.NotificationRepository;
import com.smartexpense.android.domain.repository.ResultCallback;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class NotificationRepositoryImpl implements NotificationRepository {

    private final NotificationApi api;

    public NotificationRepositoryImpl(NotificationApi api) {
        this.api = api;
    }

    @Override
    public void getNotifications(ResultCallback<List<NotificationResponseDto>> callback) {
        api.getNotifications().enqueue(new Callback<List<NotificationResponseDto>>() {
            @Override
            public void onResponse(Call<List<NotificationResponseDto>> call, Response<List<NotificationResponseDto>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body());
                } else {
                    callback.onError("Lỗi tải thông báo: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<List<NotificationResponseDto>> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }

    @Override
    public void markAsRead(String id, ResultCallback<Void> callback) {
        api.markAsRead(id).enqueue(new Callback<Void>() {
            @Override
            public void onResponse(Call<Void> call, Response<Void> response) {
                if (response.isSuccessful()) {
                    callback.onSuccess(null);
                } else {
                    callback.onError("Lỗi đánh dấu đã đọc: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<Void> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }

    @Override
    public void markAllAsRead(ResultCallback<Void> callback) {
        api.markAllAsRead().enqueue(new Callback<Void>() {
            @Override
            public void onResponse(Call<Void> call, Response<Void> response) {
                if (response.isSuccessful()) {
                    callback.onSuccess(null);
                } else {
                    callback.onError("Lỗi đánh dấu tất cả đã đọc: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<Void> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }
}

File: android/app/src/main/kotlin/com/smartexpense/android/data/repository/NotificationRepositoryImpl.kt
package com.smartexpense.android.data.repository

import com.smartexpense.android.data.remote.api.NotificationApi
import com.smartexpense.android.data.remote.dto.response.NotificationResponseDto
import com.smartexpense.android.domain.repository.NotificationRepository

class NotificationRepositoryImpl(private val api: NotificationApi) : NotificationRepository {
    override suspend fun getNotifications(): Result<List<NotificationResponseDto>> {
        return try {
            val response = api.getNotifications()
            Result.success(response)
        } catch (e: Exception) {
            Result.failure(Exception("Lỗi tải thông báo: ${e.message}", e))
        }
    }

    override suspend fun markAsRead(id: String): Result<Unit> {
        return try {
            api.markAsRead(id)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(Exception("Lỗi đánh dấu đã đọc: ${e.message}", e))
        }
    }

    override suspend fun markAllAsRead(): Result<Unit> {
        return try {
            api.markAllAsRead()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(Exception("Lỗi đánh dấu tất cả đã đọc: ${e.message}", e))
        }
    }
}
